File: aws/CloudFormation/CogCondPreAuthSettings/cogCondPreAuthSettings.go
package main

import (
	"context"
	"encoding/json"
	"github.com/aws/aws-lambda-go/cfn"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/aws/awserr"
	"github.com/aws/aws-sdk-go-v2/service/ssm"
	"github.com/codesmith-gmbh/cgc/cgcaws"
	"github.com/codesmith-gmbh/forge/aws/common"
	"github.com/mitchellh/mapstructure"
	"github.com/pkg/errors"
	"strconv"
)

func main() {
	cfg := cgcaws.MustConfig()
	p := newProc(cfg)
	lambda.Start(cfn.LambdaWrap(p.processEvent))
}

type proc struct {
	ssm *ssm.Client
}

func newProc(cfg aws.Config) *proc {
	return &proc{ssm: ssm.New(cfg)}
}

type Properties struct {
	UserPoolId       string
	UserPoolClientId string
	All              string
	Domains          []string
	Emails           []string
}

func validateProperties(input map[string]interface{}) (Properties, error) {
	var properties Properties
	if err := mapstructure.Decode(input, &properties); err != nil {
		return properties, err
	}
	if properties.UserPoolId == "" {
		return properties, errors.New("UserPoolId is required")
	}
	if properties.UserPoolClientId == "" {
		return properties, errors.New("UserPoolClientId is required")
	}
	return properties, nil
}

func (p *proc) processEvent(ctx context.Context, event cfn.Event) (string, map[string]interface{}, error) {
	properties, err := validateProperties(event.ResourceProperties)
	if err != nil {
		return "", nil, err
	}
	switch event.RequestType {
	case cfn.RequestDelete:
		return p.deleteParameter(ctx, event.PhysicalResourceID)
	case cfn.RequestCreate, cfn.RequestUpdate:
		return p.putParameter(ctx, properties)
	default:
		return common.UnknownRequestType(event)
	}
}

func (p *proc) putParameter(ctx context.Context, properties Properties) (string, map[string]interface{}, error) {
	overwrite := true
	parameterName := common.CogCondPreAuthParameterName(properties.UserPoolId, properties.UserPoolClientId)
	all, err := strconv.ParseBool(properties.All)
	if err != nil {
		return "", nil, errors.Wrapf(err, "All must be a boolean: %s", properties.All)
	}
	data := map[string]interface{}{
		"All":     all,
		"Domains": properties.Domains,
		"Emails":  properties.Emails,
	}
	dataBytes, err := json.Marshal(data)
	if err != nil {
		return "", nil, errors.Wrapf(err, "could not marshal the parameter %s", parameterName)
	}
	dataText := string(dataBytes)
	_, err = p.ssm.PutParameterRequest(&ssm.PutParameterInput{
		Overwrite: &overwrite,
		Name:      &parameterName,
		Type:      ssm.ParameterTypeString,
		Value:     &dataText,
	}).Send(ctx)
	if err != nil {
		return "", nil, errors.Wrapf(err, "could not put the parameter %s", parameterName)
	}
	return parameterName, nil, nil
}

func (p *proc) deleteParameter(ctx context.Context, parameterName string) (string, map[string]interface{}, error) {
	_, err := p.ssm.DeleteParameterRequest(&ssm.DeleteParameterInput{
		Name: &parameterName,
	}).Send(ctx)
	if err != nil {
		awsErr, ok := err.(awserr.RequestFailure)
		if !ok || awsErr.StatusCode() != 400 || awsErr.Code() != "ParameterNotFound" {
			return parameterName, nil, errors.Wrapf(err, "could not delete the parameter %s", parameterName)
		}
	}
	return parameterName, nil, nil
}

File: aws/CloudFormation/CogCondPreAuthSettings/cog_cond_pre_auth_settings.py
import json
import logging

import boto3
from box import Box
from crhelper import CfnResource
from schema import And, Optional, Schema

import aws.common.naming as naming
from aws.common.cfn import physical_resource_id, resource_properties
from aws.common.schema import not_empty

helper = CfnResource()
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

properties_schema = Schema({
    'UserPoolId': And(str, not_empty, error='not empty string for UserPoolId'),
    'UserPoolClientId': And(str, not_empty, error='not empty string for UserPoolClientId'),
    Optional('All', default=False): bool,
    Optional('Domains', default=[]): [str],
    Optional('Emails', default=[]): [str]
})

ssm = boto3.client('ssm')


def validate_properties(props):
    return Box(properties_schema.validate(props), camel_killer_box=True)


@helper.create
@helper.update
def create(event, _):
    p = validate_properties(resource_properties(event))
    parameter_name = naming.cog_cond_pre_auth_parameter_name(p.user_pool_id, p.user_pool_client_id)
    parameter_value = json.dumps({'All': p.all, 'Domains': p.domains, 'Emails': p.emails}),
    try:
        ssm.put_parameter(
            Name=parameter_name,
            Value=parameter_value,
            Overwrite=True,
            Type='String',
            Tier='Standard'
        )
    except ssm.exceptions.ClientError as e:
        raise RuntimeError(f'Cannot create parameter with name {parameter_name}') from e
    return parameter_name


@helper.delete
def delete(event, _):
    parameter_name = physical_resource_id(event)
    try:
        ssm.delete_parameter(Name=parameter_name)
    except ssm.exceptions.ParameterNotFound:
        pass
    return parameter_name


def handler(event, context):
    logger.info('event: %s', event)
    helper(event, context)
